### DISCUSSION TITLE
Add a built-in component for async view loading

### DISCUSSION THREAD

**Discussion by dead-claudia** [MithrilJS/mithril.js#2936] [Ideas]
<details>
<summary>Updates</summary>

- Switched to `signal` as that's what all the various DOM APIs use (and care about) and revised the examples (fixing a bug in the process).
- This is now a sub-bug of #2284, not #2278.
- Fix to reference `Async` instead of `m.async`.
- Update with a new bit using https://github.com/MithrilJS/mithril.js/issues/2074.
- Add `onAbort` callback to `init` for cancelling requests and similar.
- Fix to rebuild the tree on state change, rather than rebuild
</details>

I'm thinking we could add a new component for async view loading, for things like lazy component initialization based on app state, database loading, and reducing the need for route resolvers (which I'm proposing on partially removing in #2278).

Here's my proposed API:

```js
m(Async, {
	// Use this to generate the promise data.
	init: function (signal) {
		const request = requestSomeData()
		signal.onabort = () => request.cancel()
		return request.promise()
	},

	// This returns the view to render while the promise is pending.
	pending: function () {
		// ...
	},

	// This returns the view to render with the resolved promise data, in case
	// of resolution
	ready: function (resolvedData) {
		// ...
	},

	// This returns the view to render with any rejected promise error, in case
	// of rejection
	error: function (rejectedError) {
		// ...
	},
})
```

If you want to reinitialize it, just wrap it in a keyed fragment and pass a new key.

<details>
<summary>Implementation-wise, it's pretty simple, and could be implemented trivially in userland.</summary>

This, of course, is slightly more complicated, but this has the semantics I'd prefer to see mod sync handling of non-promise `return`/`throw`.

Note that this depends on https://github.com/MithrilJS/mithril.js/issues/2074#issuecomment-438452765.

```js
// v2
function Async(v) {
	let method = "pending"
	const ctrl = new AbortController()
	let data
	new Promise(resolve => resolve(v.attrs.init(ctrl.signal))).then(
		d => { method = "ready"; data = d; m.redraw() },
		e => { method = "error"; data = e; m.redraw() }
	)
	return {
		view(v) {
			let view = v.attrs[method](
				method === "pending" ? undefined : data
			)
			if (!Array.isArray(view)) view = [view]
			return m.fragment({key: method}, view)
		},
		onremove() {
			if (method === "pending") ctrl.abort()
		},
	}
}

// After #2689
function Async(v) {
	let method = "pending"
	const ctrl = new AbortController()
	let data
	new Promise(resolve => resolve(v.attrs.init(ctrl.signal))).then(
		d => { method = "ready"; data = d; m.redraw() },
		e => { method = "error"; data = e; m.redraw() }
	)
	return v => {
		let view = v.attrs[method](
			method === "pending" ? undefined : data
		)
		if (!Array.isArray(view)) view = [view]
		return m.fragment({
			key: method,
			afterRemove: () => { if (method === "pending") ctrl.abort() }
		}, view)
	}
}

// After #2278 + #2295
function Async(v, o, [method, data] = ["pending", undefined], update) {
	if (o == null) {
		const ctrl = new AbortController()
		new Promise(resolve => resolve(v.attrs.init(ctrl.signal))).then(
			d => update(["ready", d]),
			e => update(["error", e])
		)
		data = () => ctrl.abort()
	}
	return {
		next: [method, data],
		onremove: method === "pending" ? data : undefined,
		view: m(m.keyed, m(m.fragment, {key: method}, v.attrs[method](
			method === "pending" ? undefined : data
		))),
	}
}
```
</details>

**Comment by StephanHoyer** [comment]
On the first thought: I see this in userland.

**Comment by barneycarroll** [comment]
> I see this in userland.

I agree. Promises can be difficult to deal with declaratively, but ultimately this isn't something that involves any Mithril-specific logic, other than a component to store state. I wrote this component to demonstrate the potential of nested vtree functions - [`Promiser`](https://github.com/barneycarroll/proponents/blob/master/src/packages/Promiser/index.mjs) - which is a simple component that consumes a promise and exposes an explosion of promise state to the consumed view function. [Little demo here, line 39 onwards](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJHWGABwDoArBAGhAGMB7NAFxibxgA8i6AnRgAgAmMMBgCuUfkKgYAnnwC8APj4Q0RUfyUAdNHz180MAO58ACtzpYIcGAApu8OlABuMBcuA79+m4wAqEDh0Gra2AJTufJ663voOcE6utqrqjGFesQC+lIIw0jLpMXqZYSDUNrA0jBAMCIggAAyIAEwAnCDZ6Ni49QCydAIYUORUtAzMrPX0aHD81TQA1gp84ZHmltYwJPGJdoU6HFy8uSLi-NF6zhDGiCvANAAW0AIOaCXuGXpYtgDkJD98ADUfAARgADT76EEYRYAcwsojQAj4AGIwOjIXoQXR2ABaOAPDACOgmFENckARj4DWpfGaMCwfBIzQArAzMXwiESBKpYd4KeyivouHAINUGPpIOwYAIOYw6ERYvoWQ0AKQc2BgfhKlXqoV6RjcDAzMA8Rl8Q3GggYZi2XG6nL2tWFJVYDB4owQASMB56VoADmcRg5brxDxgEFhD34AecDzlRpmYpquhIABY4HwYBgbLjVLjgowOQqYWK5PoGpCwTkLrEGDQHDaYLdbMBiVhbsBZjJYDlVIZuAAJCNRxiZd5KKIcvTd2AkEs0MvLCk6acWiCLcIkH0sWxYbYyo1GF1ZSir-V8Bgg4Q8GAOLB0VwttuWCfKQwmdZWGz2RwuNyTrWSrtiQs5bAuS7yNS55KnoIHcgAoq4TAADLWBMd6-JaSbioYSI-DkOz-ie3glJC2Qco8zyvJCYRnm8ZT4HkMBVCmdQgBSaaIBS7SdCAmA4HgADqGBineIyMdMEyMGwnA8JIwhiBIKwRIBGTOBg3B8FAqhLFBH58L0GBEDBfAOIwojcLoQF8Fcxh8LcziRLYa7OCQNqGnAJA6WgCwcgA-HwrkkFRUAvCwHKIMFoXhWgyANAAuiFTxha8tg+QsES0fRkL1o2zC3EQd6irM9HeAwohEIMBWcsV6E5UKDD3o+za2ZENl6BAYC2G5HncF5GUkbEvWMJ53m6SQUgwLabntkNFFoLlaDXqaDjNU+KkfBeX6bO5UBQC5F56MgJCnRlJAaVAojwOESVukQtg0EMUAAPIAEIAEpDXoJ6ZDof2LWg0yzHwT37e9H3LDmMhA21akxEIIKiLCsJ3hkGQYEYon8DtNh7QdyDTiNnmQgAZKTtnuaN-XzstN5rQyLVkxTxM01e9N3ozrghc9PWgYwTY5M4dGmcN-NNrTK23utMDM5TswS+zq2cw+3NgwdbkK8wQsizECUnhk+4OC8mPhP957A-wkBIhDyxOfDehGClbh87AaCwj6fAAD5e3wACEfMC7CksczL3u+6zmbk5TfVeUr0tczAYRDfblOqArQNuD7lMxa84e2fFCWmeZlm6M45uA5btX9ehdvtRkVcC9wqOaHw1sCBDPX7DEXV87H42+SnVNjedvi2E3LeTZYOQT9NhSZIxFQsbh7EAMwUtxaYdJQXSCfUACSaA+VsFCSeMLAyfUhzyScSlW4irEMF3tZV2AyyHUqbm5ywhdBUK5NyhkEVOgb8v4pVioXZK1EWAKHkFBH4YAH64R+MzDkYDoFxUSlA1KEUYjdz0CXKyKwORgEhP5DkH9YJ6DACQIgFh5SMCAW4aONC6F0AYUwi61xgxHT4OQ3hegDLdVOu5ZuogcBMDgN9fQUUBFt1sCIzSsJxEXykRyIasilQdX0HZIwz81wEOmqXYhcjvCRy4cYAx+h+FUM-sPGmuiFGnSUSoyR0ilSaNsWY+xXlv6Ay8e4vgAMlTBL4PPbeu8eggBeuiCS1ApIXzwFXIqNcQZQXth4DIL0QRkGXiQBYMAZBwD7tTKRJBVoIRhA8ceTD65Ck1gLZgJAYDIUYF5bkNSir4P0J6JEJJ3ICAEEhC+aFZgsEwowoqQtxbawtNwa6v154W1qHMbM3BiRGF0Okupx0RENIli01RSUKlVM6QBTJQpekbIPKrGAwzULoXGdwM50ytYwBnvMpOGQyJvHPNfY4QhTjKQyXcdS3CHKbW0PU7BsUGp6AYKoMUELVgOx0TMrYhzJHLAMgAZWmt8uFl4gb5TcIVOqpUMgVSqk2CFKSSqMDKkSsOtxmCaQ2WgQlhIkSwHuYwW45ypxCl7lsSZbhVAx1KXwaOftZqWBCuMUSMxbAiqUXPMI2jI7IBFUwhKyruneGFQeE2Jg-ZwLbkMGwgSjaHlNvNAlpl44OEqtVVq0hmBpIuWY5Ybr4BFhiAtEoET+LdDwLjcSp94nn0mCAf5CkgXnDBTcOGihIToJwZgpKfjWx7PRSUQlCK0BiluNDWGIKbLZNyVUdycBRSwjQHzN5NY0FDGurcHRdAvTUkZd4O8FhuBtsuB25EDRu1xD-K4AQtwRBQBsKOwxlbmCTvNTO95HIipIl5G2w010518F8IwWAS7l2zvIvquZcgNXooui2gCfBMbYwlWNNh35ZYXn2U0oiE7ljbtfaRSET1GCPD1Zet5zTuB9uWL+sWoGHALplN+z55FITW2ehetBV7108ndssadNh0Ogf3YehD11Rb6Gtcas2QpQkLUpXTZWzqmwtjQCSGe4Y0CqU9foXuzGjA+NoRYF9-szU7jQHx59mxpE8e3BgYOBaxS2B479eiC9yjMUfjMPAq8uIUiDQJKJcSxhMESfUQIRx+AEBgMQbwYABN8D+AAegs8QFBgNTM3xEmJLS1DbMOY88wbgLm3PHEPsfSUPmSD2ZC6oGAgWsBmb4DEt+sQbOWDsxFxLsX4thq0ilxkDnssuZ0EF-g-RBhQGs+F+zpWhiFaBisvgABBIgiplh-ABMCcEGRTRMFxCIKwUA5BwCtPiO8XUMhckGbyWIAosA6AhIDWA-AZMOHg1BXDr6dD7gfIiRgthiQ0FcYwEg2IBAyCbTEXRLYON-28N8JrRAcghl+A8Ck7XQRgmYOwRguIhiRl0DQC+d5qx2eHPtOgPxdZKgoVYv2y3LOyl4TY2x3w-N3hyBlSIVivi2ES+drx3gaA6UWPynZ+PvBw9W3Mkjpj9ALTJ9kG7ZO+Ao+xmjqIGUGdY9iN8arUBHs06VD8AAwkTpYhZRRCEvG-H01hmcDCGBaOgoNxiqGugRLn3hIdM61wE3dmu1yeKod8KLhgcgos4145JAnNgKA18zg8ABHa6sxnujSIHARA9n7MUAYEQaQAOHhOCENwbcQCNwDC2PQLA9mRRtPs69wJVDtxsdsE5h6ApV7Jzt1zwh1k7eXchSmgXZHbDZZrOJmwDPFSTkoUz2IpBMO8jt94JHdfbu-CIG98En3vu-braDQH3BqzN9gj8FCdAiS8hyH77MNg+BYzFCIlzbfYiJ9sYblfnIwN9pH3oQKu-9CE2L0bjvXf5ub9gj3n7Ol+8A6M1pA-Sp6BQB4EqFENBySP-0MP4-tifgAFUHgCwHgSA+AcVLBponhsMjAL558LB3ZEBl8L99BdYv9scfg6EYsgR3s0CoQYQFh4RggkRUQwAaAaBcC9AJssM+R9AZsKDQQeBg9Yh39P9f8lQf9kD69t8eBcC18vEi42CHIv8j9OD0CHhmgCIt9LprptwxRYBUDBDvBSBpCtgTsZBQI-d5N7MtAdB7MwgSB7pbAuQjR4RjJfRJx6DvgMDJDjDlszC+D8cddN8BCV9k49csgrEfklQnD55ShVNKgV48B-QmgKRmhcRV5-REAGgOgEpqAMo6hUBg094QArAfRuBoBGJLIoA8BoxGB3dPd7NEQiACD5Vo9UiHh0ioAAABCkEgWogANns3KMqIM1FTwDgAbAgCIBkj4j0zwBBBrUyO4GyPqFyPyK9yKJKKj3swGLgBqLqJIBZBmJrVaKYXaM6O6JiMyCAA). IMO this is a lot neater - less OO, more obvious state exposure - but I don't think it belongs in core.

**Comment by dead-claudia** [comment]
It's not really *mandatory* to have in core, and I addressed the primary use of route resolvers (auth) with a much better-suited abstraction. Would you all be okay with an out-of-core-but-officially-blessed `mithril/async`?

This is probably the least significant facet of #2278, and I wouldn't have an issue separating this from that.

**Comment by StephanHoyer** [comment]
not bundled in core but opt in module would be great thought

**Comment by dead-claudia** [comment]
@StephanHoyer Yeah, I'm starting to agree with that thought. I'll remove it from my list in #2278 and just keep it with my (newly filed) #2284.
